packages/babel-parser/src/parse-error.js

```diff
@@ -0,0 +1,211 @@
+// @flow
+
+import { Position } from "./util/location";
+import type { NodeBase } from "./types";
+import {
+  type ParseErrorCode,
+  ParseErrorCodes,
+  type ParseErrorCredentials,
+} from "./parse-error/credentials";
+
+const ArrayIsArray = Array.isArray;
+const {
+  assign: ObjectAssign,
+  defineProperty: ObjectDefineProperty,
+  getPrototypeOf: ObjectGetPrototypeOf,
+  keys: ObjectKeys,
+} = Object;
+
+type ToMessage<ErrorDetails> = (self: ErrorDetails) => string;
+
+const StandardMessage = Symbol("StandardMessage");
+
+// This should really be an abstract class, but that concept doesn't exist in
+// Flow, outside of just creating an interface, but then you can't specify that
+// it's a subclass of `SyntaxError`. You could do something like:
+//
+// interface IParseError<ErrorDetails> { ... }
+// type ParseError<ErrorDetails> = SyntaxError & IParseError<ErrorDetails>;
+//
+// But this is just a more complicated way of getting the same behavior, with
+// the added clumsiness of not being able to extends ParseError directly. So,
+// to keep things simple and prepare for a Typescript future, we just make it a
+// "private" superclass that we exclusively subclass:
+
+export class ParseError<ErrorDetails> extends SyntaxError {
+  code: ParseErrorCode;
+  reasonCode: string;
+
+  loc: Position;
+  details: ErrorDetails;
+
+  // There are no optional fields in classes in Flow, so we can't list these
+  // here: https://github.com/facebook/flow/issues/2859
+  // syntaxPlugin?: SyntaxPlugin
+  // missingPlugin?: string[]
+
+  constructor({
+    loc,
+    details,
+  }: {
+    loc: Position,
+    details: ErrorDetails,
+  }): ParseError<ErrorDetails> {
+    super();
+
+    this.loc = loc;
+
+    ObjectDefineProperty(this, "details", {
+      value: details,
+      enumerable: false,
+    });
+
+    // $FlowIgnore
+    if (details.missingPlugin) {
+      ObjectDefineProperty(this, "missingPlugin", {
+        get() {
+          return this.details.missingPlugin;
+        },
+        enumerable: true,
+      });
+    }
+
+    return this;
+  }
+
+  clone({
+    loc,
+    details,
+  }: {
+    loc?: Position,
+    details?: ErrorDetails,
+  } = {}) {
+    return new (ObjectGetPrototypeOf(this).constructor)({
+      loc: loc || this.loc,
+      details: { ...this.details, ...details },
+    });
+  }
+
+  get pos() {
+    return this.loc.index;
+  }
+}
+
+function toParseErrorClass<ErrorDetails>(
+  toMessage: ToMessage<ErrorDetails>,
+  credentials: ParseErrorCredentials,
+): Class<ParseError<ErrorDetails>> {
+  return class extends ParseError<ErrorDetails> {
+    #message: typeof StandardMessage | string = StandardMessage;
+
+    constructor(...args): ParseError<ErrorDetails> {
+      super(...args);
+      // $FlowIgnore - Only necessary because we can't make syntaxPlugin optional.
+      ObjectAssign(this, credentials);
```

Comment:
I don't understand what's happening here. Regardless of what the flow type say, `toParseErrorCredentials` returns a two-elements array. Are we assigning `this[0]` and `this[1]`? Why do we use indexes rather than property names? Are then these properties read somewhere?

Reply:
Credentials isn't an array, its a `ParseErrorCredentials` (in the type declaration above):

```javascript
function toParseErrorClass<ErrorDetails>(
  toMessage: ToMessage<ErrorDetails>,
  credentials: ParseErrorCredentials,
)
``` 

And `ParseErrorCredentials` is defined as:

```javascript
export type ParseErrorCredentials = {
  code: ParseErrorCode,
  reasonCode: string,
  syntaxPlugin?: SyntaxPlugin,
};
```

Thus, if we could *also* make `syntaxPlugin` an optional field of `ParseError`, then Flow wouldn't complain. But since `syntaxPlugin` is *not present* in `ParseError` at all (because that would force errors with no `syntaxPlugin` to have a `syntaxPlugin` field with `undefined`), we can't call `Object.assign` without Flow saying "wait a minute, this MAY have a syntaxPlugin, which ParseError *doesn't* have!" So, this is equivalent to the following code:

```javascript
      // Items everyone agrees we have:
      const { syntaxPlugin = NoPresent, ...undisputed } = credentials;
      ObjectAssign(this, undisputed);
      // No-op but makes Flow happy
      if (syntaxPlugin != NotPresent) {
          this.syntaxPlugin = syntaxPlugin;      
      }
```